mne/tests/test_epochs.py

```diff
@@ -745,6 +745,18 @@ def test_epochs_baseline(preload):
     expected[0] = [-0.5, 0.5]
     assert_allclose(epochs.get_data()[0], expected)
 
+    # Test that Epochs.baseline is only updated if the data was actually
+    # rescalrd.
+    baseline = (None, None)
+    epochs = mne.Epochs(raw, events, None, 0, 1e-3, baseline=baseline,
+                        preload=preload)
+    epochs.apply_baseline(None)
+
+    if preload:
+        assert epochs.baseline == baseline
+    else:
+        assert epochs.baseline is None
```

Comment:
I don't understand why the data rescaling depends on whether or not `preload` was used. It seems more like in the non-preload case that there should be some "pending baseline operation" that gets applied after data are loaded from disk so that these actually match.

In other words, after `epochs.apply_baseline(...)`, both the `epochs.baseline` attribute and the result of `epochs.get_data()` should be the same when `preload=True` as when `preload=False`.

Reply:
> I don't understand why the data rescaling depends on whether or not `preload` was used. It seems more like in the non-preload case that there should be some "pending baseline operation" that gets applied after data are loaded from disk so that these actually match.
> 
> In other words, after `epochs.apply_baseline(...)`, both the `epochs.baseline` attribute and the result of `epochs.get_data()` should be the same when `preload=True` as when `preload=False`.

Yeah that's what I thought at first too, until I figured that wasn't the case :) I used the following MWE do investigate the issue on `master`:

```python
# %%
import mne
import os
import numpy as np

preload = True
sample_data_folder = mne.datasets.sample.data_path()
sample_data_raw_file = os.path.join(sample_data_folder, 'MEG', 'sample',
                                    'sample_audvis_raw.fif')
raw = mne.io.read_raw_fif(sample_data_raw_file, verbose=False).crop(tmax=60)
events = mne.find_events(raw, stim_channel='STI 014', verbose=False)

# %%
baseline = (None, 0)
baseline2 = None

# Apply a baseline
epochs = mne.Epochs(raw, events, tmin=-0.3, tmax=0.7, baseline=baseline,
                    preload=preload, verbose=False)

# Remove the baseline
epochs2 = epochs.copy().apply_baseline(baseline2)

print(epochs.baseline)
print(epochs2.baseline)

# %%
if np.allclose(epochs.get_data(),
               epochs2.get_data()):
    print('Data is identical')
else:
    print('Data differs')
```
## `preload=True` on `master`
When we use `preload=True`, the following output it produced:

```
(None, 0)
None
Data is identical
```
i.e., although the values in `.baseline` differ, the data is the same. We can find out why this is the case by looking at `Epochs.apply_baseline`:
https://github.com/mne-tools/mne-python/blob/ca3e787ca40059ee72b93e517adbe7506688c2a3/mne/epochs.py#L661-L671
When we create `epochs` in our code:
```python
epochs = mne.Epochs(raw, events, tmin=-0.3, tmax=0.7, baseline=baseline,
                    preload=True, verbose=False)
```
we enter the `if` branch, and the call to `rescale()` will actually rescale the data according to our baseline `(None, 0)`. After that, the `.baseline` attribute will be set.

Now when we apply the "baseline" `None` to `epochs2`:
 ```python
epochs2 = epochs.copy().apply_baseline(baseline2)
```
we will follow the same branching as in the former case, **but** `rescale` won't actually do anything, as when `baseline is None` it will leave the input data unmodified:
https://github.com/mne-tools/mne-python/blob/ca3e787ca40059ee72b93e517adbe7506688c2a3/mne/baseline.py#L62-L66
Then, like in the first case, the `.baseline` attribute will be set to the new value.

**This means that in this case, even though no rescaling took place, the `.baseline` attribute of the preloaded data was altered.**

## `preload=False` on `master`
Let's test with `preload=False`:
```
(None, 0)
None
Data differs
```
So the `.baseline` attribute was updated, just like in the previous example. But `get_data()` now yields different results. Why? Because if we don't preload the data, obviously a the to `apply_baseline()` won't perform any rescaling; instead, it **only** updates the `.baseline` attribute. **When** we finally load the data, this baseline operation is carried out: a baseline of `(None, 0)` is applied to `epochs`, and no baseline correction (`baseline=None`) is applied to `epochs2`.

## Summary for `master`
- When using `preload=True`, followed by `apply_baseline(None)`, we may end up with an `Evoked` that has `Evoked.baseline = None` set, even though a baseline had been applied earlier. **I consider this a bug.**
- When using `preload=False`, `apply_baseline()` may be used to toggle on / off baseline correction until we finally decide to call `get_data()`. **I consider this expected behavior.**


## Behavior in this PR branch

### `preload=True`
```
(None, 0)
(None, 0)
Data is identical
```
The rescaling no-op of
```python
epochs2 = epochs.copy().apply_baseline(baseline2)
```
correctly did **not** alter the `.baseline` attribute, as the preloaded data was not altered.

### `preload=False`
```
(None, 0)
None
Data differs
```
Behavior here is identical to `master`: As long as the data has not been preloaded, baseline correction can be toggled, and will be applied upon data retrieval via `get_data()`.

## Summary for this PR branch
- When using `preload=True`, followed by `apply_baseline(None)`, `Evoked.baseline` stays unchanged, which correctly reflects that the data has not been altered by the `apply_baseline()` call.  **I consider this expected behavior.**
- Same behavior as on `master`: When using `preload=False`, `apply_baseline()` may be used to toggle on / off baseline correction until we finally decide to call `get_data()`. **I consider this expected behavior.**

## Summary
I believe I'm doing the Right Thing™️ in this PR.

Reply:
There's still a UX issue IMHO if one
- reads a baseline-corrected Evoked or Epochs **without preloading** from disk
- then invokes `evoked.apply_baseline(None)`
- only after that loads the data

this should **not** be possible, as we cannot remove an existing baseline correction, as @agramfort pointed out in https://github.com/mne-tools/mne-python/pull/8156#discussion_r487387001

I guess we'll have to re-think the logic of this.

Reply:
… for example, what if you

1. read **not-baseline-corrected** Evoked without preloading from disk
1. then invoke `evoked.apply_baseline((None, 0))`
1. then wishes to invoke evoked.apply_baseline(None)
1. and only after that loads the data

At step number 3, we wouldn't be able to tell whether the none-preloaded data with a `baseline != None` had been read from disk as a baseline-corrected dataset, or whether what happened was what I show in this example

Reply:
We could simply **drop support** for `apply_baseline(None)`, this would make these problems go away.
Opinions? @agramfort?

Reply:
> … for example, what if you

If it simplifies things, I would raise an error on step (3). Even though in theory it's possible for us to do what the users are asking for, the fact that it depends on `preload` or not is not good. Basically we want to discourage people from doing `.apply_baseline((None, 0)).apply_baseline(None)`, even if in some circumstances it should technically be possible to do so.

> We could simply drop support

-1 on this, in pipelines it's nice to be able to set a var at the top and then just do `apply_baseline(baseline)` and this would require people to modify their code to say `if baseline is not None:` instead.

I would just always prevent going from a `.apply_baseline(non_None)`ed data to a `.apply_baseline(None)`ed data, then we have consistent behavior that we can tell people, and hopefully it simplifies the code end.

Reply:
> I would just always prevent going from a `.apply_baseline(non_None)`ed data to a `.apply_baseline(None)`ed data, then we have consistent behavior that we can tell people, and hopefully it simplifies the code end.

Sounds good, will implemenent